README.rst
Flask-Language
==============

|PyPI Version| |PyPI License| |PyPI Versions| |Build Status| |Coverage
Status| |Documentation Status|

Flask-Language is a Flask_ extension providing a
simple mechanism to handle a client-side language cookie.

It is somewhat loosely based on this snippet: `<http://flask.pocoo.org/snippets/128>`__

.. _Flask: http://flask.pocoo.org

Installation
------------

Install the extension with with pipenv_ (recommended):

::

    $ pipenv install flask-language

Or with pip_:

::

    $ pip install flask-language

.. _pip: https://pip.pypa.io
.. _pipenv: https://docs.pipenv.org

Usage
-----

Initialize the Flask-Language extension (also compatible with the `Application Factories`_ pattern):

::

        from flask import Flask, jsonify
        from flask_language import Language, current_language

        app = Flask(__name__)
        lang = Language(app)

Define the language hooks:

::

        @lang.allowed_languages
        def get_allowed_languages():
            return ['en', 'fr']

        @lang.default_language
        def get_default_language():
            return 'en'

Define the desired end-points to retrieve and manipulate the current language:

::

        @app.route('/api/language')
        def get_language():
            return jsonify({
                'language': str(current_language),
            })

        @app.route('/api/language', methods=['POST'])
        def set_language():
            req = request.get_json()
            language = req.get('language', None)

            lang.change_language(language)

            return jsonify({
                'language': str(current_language),
            })

Before each request, Flask-Language will automatically determine the current
language in the following order:

1. The language cookie (if any and matching the allowed languages)
2. The ``Accept-Language`` HTTP header (if any and matching the allowed languages)
3. The provided default language

During each request context, the current language can be accessed using
``current_language``.

After each request, the current language will be stored in the language cookie.

.. _Application Factories: http://flask.pocoo.org/docs/0.12/patterns/appfactories/

Configuration
-------------

Flask-Language is configurable via the following configuration variables:

- ``LANGUAGE_COOKIE_NAME``: name for the cookie language (default: ``'lang'``)
- ``LANGUAGE_COOKIE_TIMEOUT``: validity duration of the cookie language (default: ``datetime.timedelta(days=365)``)
- ``LANGUAGE_COOKIE_DOMAIN``: domain for the cookie language (default: ``None``)
- ``LANGUAGE_COOKIE_SECURE``: set secure option for the cookie language (default: ``False``)
- ``LANGUAGE_COOKIE_HTTPONLY``: set HTTP-only for the cookie language (default: ``False``)

Documentation
-------------

The Sphinx-compiled documentation is available on
`ReadTheDocs <http://flask-language.readthedocs.io/en/latest/>`__.

License
-------

The MIT License (MIT)

Copyright (c) 2018 Romain Clement

.. |PyPI Version| image:: https://img.shields.io/pypi/v/flask-language.svg
   :target: https://pypi.python.org/pypi/flask-language
.. |PyPI License| image:: https://img.shields.io/pypi/l/flask-language.svg
   :target: https://pypi.python.org/pypi/flask-language
.. |PyPI Versions| image:: https://img.shields.io/pypi/pyversions/flask-language.svg
   :target: https://pypi.python.org/pypi/flask-language
.. |Build Status| image:: https://travis-ci.org/rclement/flask-language.svg?branch=master
   :target: https://travis-ci.org/rclement/flask-language
.. |Coverage Status| image:: https://coveralls.io/repos/github/rclement/flask-language/badge.svg?branch=master
   :target: https://coveralls.io/github/rclement/flask-language?branch=master
.. |Documentation Status| image:: https://readthedocs.org/projects/flask-language/badge/?version=master
   :target: http://flask-language.readthedocs.io/en/master/


HISTORY.rst
Release History
===============

0.1.0 (2018-03-07)
------------------

- Initial release of Flask-Language


tests/test_language.py
import unittest
import json

try:
    from http.cookies import SimpleCookie
except ImportError:
    from Cookie import SimpleCookie

from flask import Flask, jsonify, request
from flask_language import Language, current_language


# ------------------------------------------------------------------------------


class JSONRequest:
    def __init__(self, data):
        self.json = json.dumps(data)
        self.content_type = 'application/json'


def get_json_data(response):
    return json.loads(response.data)


def get_cookies(response):
    return SimpleCookie(response.headers.get('set-cookie'))


# ------------------------------------------------------------------------------


class TestLanguage(unittest.TestCase):

    def assert_language_response(self, response, lang):
        cookies = get_cookies(response)
        cookie_lang = cookies.get(self.lang._cookie_name)
        self.assertNotEqual(cookie_lang, None)
        self.assertEqual(cookie_lang.value, lang)

        data = get_json_data(response)
        self.assertEqual(data.get('language'), lang)

    def setUp(self):
        app = Flask(__name__)
        app.testing = True
        self.app = app
        self.client = app.test_client(use_cookies=True)
        self.lang = Language(app)
        self.allowed_languages = ['en', 'fr']
        self.default_language = 'en'

        @self.lang.allowed_languages
        def allowed_languages():
            return self.allowed_languages

        @self.lang.default_language
        def default_language():
            return self.default_language

        @self.app.route('/language')
        def get_language():
            return jsonify({
                'language': str(current_language)
            }), 200

        @self.app.route('/language', methods=['POST'])
        def set_language():
            req = request.get_json()
            language = req.get('language', None)

            self.lang.change_language(language)

            return jsonify({
                'language': str(current_language),
            }), 201

    def test_get_default_language(self):
        rv = self.client.get('/language')
        self.assertEqual(rv.status_code, 200)
        self.assertEqual(rv.mimetype, 'application/json')
        self.assert_language_response(rv, self.default_language)

    def test_change_allowed_language(self):
        new_lang = 'fr'

        rq = JSONRequest({
            'language': new_lang
        })

        rv = self.client.post(
            '/language', data=rq.json, content_type=rq.content_type
        )
        self.assertEqual(rv.status_code, 201)
        self.assertEqual(rv.mimetype, 'application/json')
        self.assert_language_response(rv, new_lang)

    def test_change_not_allowed_language(self):
        new_lang = 'dk'

        rq = JSONRequest({
            'language': new_lang
        })

        rv = self.client.post(
            '/language', data=rq.json, content_type=rq.content_type
        )
        self.assertEqual(rv.status_code, 201)
        self.assertEqual(rv.mimetype, 'application/json')
        self.assert_language_response(rv, self.default_language)


if __name__ == '__main__':
    unittest.main()


flask_language.py
'''
    Flask-Language
    --------------

    Flask-Language is a Flask extension providing a simple mechanism to handle
    a client-side language cookie.

    It is somewhat loosely based on this snippet:
    http://flask.pocoo.org/snippets/128

    :copyright: (c) 2018 by Romain Clement.
    :license: MIT, see LICENSE for more details.
'''


from datetime import timedelta

from flask import _request_ctx_stack as stack, request
from werkzeug.local import LocalProxy


__all__ = ['Language', 'current_language']


_ctx_err_msg = '''
Working outside of an application context.
'''


def _set_language(language):
    ctx = stack.top
    if ctx is None:
        raise RuntimeError(_ctx_err_msg)
    setattr(ctx, 'language', language)


def _find_language():
    ctx = stack.top
    if ctx is None:
        raise RuntimeError(_ctx_err_msg)
    return getattr(ctx, 'language', None)


current_language = LocalProxy(_find_language)


class Language(object):
    '''
    Primary class container for language cookie handling.

    Before each request, it will try to retrieve the current language in the
    language cookie. If no cookie is present, it will try to find a best match
    of allowed languages in the 'Accept-Language' HTTP header. Finally if still
    no language is found, it will use the provided default language.

    Each view can then access the current language using `current_language`.

    After each request, the current language will be stored in the language
    cookie.

    You might initialize :class:`Language` like this::

        language = Language()

    Then pass the application object to be configured::

        language.init_app(app)
    '''

    def __init__(self, app=None):
        self._allowed_languages = None
        self._default_language = None

        if app is not None:
            self.init_app(app)

    def init_app(self, app):
        '''
        Initializes a Flask object `app`: binds language cookie retrieval with
        app.before_request and binds language cookie installation with
        app.after_request.

        :param app: The Flask application object.
        '''

        app.before_request(self._before_request)
        app.after_request(self._after_request)

        self._cookie_name = app.config.get('LANGUAGE_COOKIE_NAME', 'lang')
        self._cookie_timeout = app.config.get('LANGUAGE_COOKIE_TIMEOUT',
                                              timedelta(days=365))
        self._cookie_domain = app.config.get('LANGUAGE_COOKIE_DOMAIN', None)
        self._cookie_secure = app.config.get('LANGUAGE_COOKIE_SECURE', False)
        self._cookie_httponly = app.config.get('LANGUAGE_COOKIE_HTTPONLY',
                                               False)

        return self

    def allowed_languages(self, callback):
        '''
        A decorator to be used to specify the list of allowed languages.

        Example usage of :class:`allowed_languages` might look like this::

            language = Language(app)

            @language.allowed_languages
            def get_allowed_languages():
                return ['en', 'fr']

        :param callback: The callback to be wrapped by the decorator.
        '''

        self._allowed_languages = callback
        return callback

    def default_language(self, callback):
        '''
        A decorator to be used to specify the default language.

        Example usage of :class:`default_language` might look like this::

            language = Language(app)

            @language.default_language
            def get_default_language():
                return 'en'

        :param callback: The callback to be wrapped by the decorator.
        '''

        self._default_language = callback
        return callback

    def change_language(self, language):
        '''
        Allows to change the current language to the specified one.
        The `current_language` will reflect the provided language only if it is
        present in :class:`allowed_languages`.

        Example usage of :class:`change_language` might look like this::

            language = Language(app)

            @app.route('/api/language', methods=['POST'])
            def set_language():
                req = request.get_json()
                lang = req.get('lang')

                language.change_language(lang)

                return jsonify({
                    'lang': str(current_language)
                })

        :param language: The new language to be set.
        '''
        allowed_languages = self._allowed_languages()
        if language in allowed_languages:
            _set_language(language)

    def _before_request(self):
        '''
        Retrieve the current language using the following policy:
            1. Try to extract the language cookie
            2. If no language cookie is found, try to find a match between
               `request.accept_languages` and :class:`allowed_languages`
            3. If no match is found, use the :class:`default_language`

        Afterwards, the `current_language` will be available.
        '''

        if self._allowed_languages is None:
            raise RuntimeError(
                '''
                Specify a callback to retrieve the allowed languages, using the
                allowed_languages decorator.
                '''
            )

        if self._default_language is None:
            raise RuntimeError(
                '''
                Specify a callback to retrieve the default language, using the
                default_language decorator.
                '''
            )

        allowed_languages = self._allowed_languages()
        default_language = self._default_language()

        cookie_lang = request.cookies.get(self._cookie_name)
        request_lang = request.accept_languages.best_match(allowed_languages)
        language = cookie_lang or request_lang or default_language

        _set_language(language)

    def _after_request(self, response):
        '''
        Retrieve the current language and store the value in the language
        cookie.

        :param response: A Flask Response object.
        '''

        lang = _find_language()
        if lang is not None:
            response.set_cookie(key=self._cookie_name,
                                value=lang,
                                max_age=self._cookie_timeout,
                                domain=self._cookie_domain,
                                secure=self._cookie_secure,
                                httponly=self._cookie_httponly)
        return response
